File: src/notion_extractor/extract/resolve.py
import os
import re
from dataclasses import dataclass
from pathlib import Path

_LINK_RE = re.compile(r"\{\{notion-link:([^|}]+)(?:\|([^}]+))?\}\}")
_ASSET_RE = re.compile(r"\{\{notion-asset:([^}]+)\}\}")

BrokenLink = tuple[str, str | None, Path]


@dataclass
class LinkContext:
    from_file: Path
    id_to_path: dict[str, Path]
    url_to_local: dict[str, Path]
    vault_root: Path


def resolve_placeholders(md: str, ctx: LinkContext) -> tuple[str, list[BrokenLink]]:
    broken: list[BrokenLink] = []

    def link_sub(match: re.Match[str]) -> str:
        notion_id = match.group(1)
        label = match.group(2)
        target_path = ctx.id_to_path.get(notion_id)
        if target_path is not None:
            page_name = target_path.stem
            return f"[[{page_name}]]"
        broken.append((notion_id, label, ctx.from_file))
        return label if label else "(link removed)"

    def asset_sub(match: re.Match[str]) -> str:
        url = match.group(1)
        local = ctx.url_to_local.get(url)
        if local is None:
            return url
        rel = os.path.relpath(local, start=ctx.from_file.parent)
        return rel.replace(os.sep, "/")

    md = _LINK_RE.sub(link_sub, md)
    md = _ASSET_RE.sub(asset_sub, md)
    return md, broken


def render_report(
    *,
    vault_root: Path,
    pages_extracted: int,
    items_extracted: int,
    attachments_downloaded: int,
    total_bytes: int,
    duration_s: float,
    broken_links: list[BrokenLink],
    failures: list[tuple[str, str]],
    warnings: list[str],
) -> str:
    mb = total_bytes / (1024 * 1024)
    lines = [
        "# Relatório de extração",
        "",
        f"- Páginas extraídas: {pages_extracted}",
        f"- Itens de database extraídos: {items_extracted}",
        f"- Anexos baixados: {attachments_downloaded} ({mb:.1f} MB)",
        f"- Duração: {duration_s:.1f}s",
        "",
    ]
    if broken_links:
        lines.append(f"## Links para fora da seleção ({len(broken_links)})")
        for notion_id, label, from_file in broken_links:
            rel = from_file.relative_to(vault_root)
            label_part = f" — {label}" if label else ""
            lines.append(f"- `{notion_id}`{label_part} (em [{rel}]({rel}))")
        lines.append("")
    if failures:
        lines.append(f"## Falhas ({len(failures)})")
        for nid, reason in failures:
            lines.append(f"- `{nid}`: {reason}")
        lines.append("")
    if warnings:
        lines.append(f"## Avisos ({len(warnings)})")
        for w in warnings:
            lines.append(f"- {w}")
        lines.append("")
    return "\n".join(lines)

File: app/src/server/extract/resolve.ts
import { posix as path } from 'node:path';

const LINK_RE = /\{\{notion-link:([^|}]+)(?:\|([^}]+))?\}\}/g;
const ASSET_RE = /\{\{notion-asset:([^}]+)\}\}/g;

export interface BrokenLink {
  id: string;
  label: string | null;
  fromFile: string;
}

export interface LinkContext {
  fromFile: string;
  idToPath: Map<string, string>;
  urlToLocal: Map<string, string>;
  vaultRoot: string;
}

export function resolvePlaceholders(
  md: string,
  ctx: LinkContext,
): { md: string; broken: BrokenLink[] } {
  const broken: BrokenLink[] = [];
  let out = md.replace(LINK_RE, (_match, id: string, label: string | undefined) => {
    const targetPath = ctx.idToPath.get(id);
    if (targetPath !== undefined) {
      const pageName = path.basename(targetPath, '.md');
      return `[[${pageName}]]`;
    }
    broken.push({ id, label: label ?? null, fromFile: ctx.fromFile });
    return label ?? '(link removed)';
  });
  out = out.replace(ASSET_RE, (_match, url: string) => {
    const local = ctx.urlToLocal.get(url);
    if (!local) return url;
    const rel = path.relative(path.dirname(ctx.fromFile), local);
    return rel;
  });
  return { md: out, broken };
}

export interface ReportInput {
  vaultRoot: string;
  pagesExtracted: number;
  itemsExtracted: number;
  attachmentsDownloaded: number;
  totalBytes: number;
  durationS: number;
  brokenLinks: BrokenLink[];
  failures: Array<{ id: string; reason: string }>;
  warnings: string[];
}

export function renderReport(input: ReportInput): string {
  const mb = input.totalBytes / (1024 * 1024);
  const lines: string[] = [
    '# Relatório de extração',
    '',
    `- Páginas extraídas: ${input.pagesExtracted}`,
    `- Itens de database extraídos: ${input.itemsExtracted}`,
    `- Anexos baixados: ${input.attachmentsDownloaded} (${mb.toFixed(1)} MB)`,
    `- Duração: ${input.durationS.toFixed(1)}s`,
    '',
  ];
  if (input.brokenLinks.length) {
    lines.push(`## Links para fora da seleção (${input.brokenLinks.length})`);
    for (const b of input.brokenLinks) {
      const rel = path.relative(input.vaultRoot, b.fromFile);
      const labelPart = b.label ? ` — ${b.label}` : '';
      lines.push(`- \`${b.id}\`${labelPart} (em [${rel}](${rel}))`);
    }
    lines.push('');
  }
  if (input.failures.length) {
    lines.push(`## Falhas (${input.failures.length})`);
    for (const f of input.failures) lines.push(`- \`${f.id}\`: ${f.reason}`);
    lines.push('');
  }
  if (input.warnings.length) {
    lines.push(`## Avisos (${input.warnings.length})`);
    for (const w of input.warnings) lines.push(`- ${w}`);
    lines.push('');
  }
  return lines.join('\n');
}
